psl/psl-core/src/validate/validation_pipeline/validations/relations.rs

```diff
@@ -525,3 +526,39 @@ fn is_empty_fields<T>(fields: Option<impl ExactSizeIterator<Item = T>>) -> bool
         Some(fields) => fields.len() == 0,
     }
 }
+
+/// There cannot be any required field in a relation where one of the referential actions is SetNull.
+pub(crate) fn required_relation_cannot_use_set_null(relation: InlineRelationWalker<'_>, ctx: &mut Context<'_>) {
+    // return early if there's no relation field on the referencing model
+    let forward = match relation.forward_relation_field() {
+        Some(forward) => forward,
+        None => return,
+    };
+
+    // return early if no referencing field is required
+    if forward
+        .referencing_fields()
+        .map(|mut fields| fields.all(|f| !f.ast_field().arity.is_required()))
+        .unwrap_or_default()
+    {
+        return;
+    }
+
+    if let Some(ReferentialAction::SetNull) = forward.explicit_on_delete() {
+        ctx.push_error(DatamodelError::new_attribute_validation_error(
+            r#"The `onDelete` referential action of a relation must not be set to `SetNull` when a referenced field is required.
+Either choose another referential action, or make the referenced fields optional."#,
+            RELATION_ATTRIBUTE_NAME,
+            forward.ast_field().span(),
+        ))
+    }
+
+    if let Some(ReferentialAction::SetNull) = forward.explicit_on_update() {
+        ctx.push_error(DatamodelError::new_attribute_validation_error(
+            r#"The `onUpdate` referential action of a relation must not be set to `SetNull` when a referenced field is required.
+Either choose another referential action, or make the referenced fields optional."#,
```

Comment:
Either use indoc macro for these string slices or put them into the same line (no line break).